Reverse invoice

lomamech opened the pull request:
Implementation for credit note,
instead reverse transaction, we have create the same transtation with the sign minous 

@jniles

lomamech committed:
Sanitaze code

jniles commented on server/models/schema.sql:
```diff
@@ -1564,6 +1564,8 @@ CREATE TABLE `invoice` (
   `created_at`          TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
   `is_distributable`    TINYINT(1) NOT NULL ,
+  `is_credit_note`      TINYINT(1),
+  `credit_note_by`      SMALLINT(5) UNSIGNED,
```
This information should be stored in the posting journal.  Why do you need it in this table?

lomamech replied:
Just because / # / invoices draws data directly from the Invoices table, which is why I added the column and finally is_credit_note credit_note_by to avoid queries joins with the posting log table but also of general_ledger in the case or posting journal is empty

jniles commented on server/models/schema.sql:
```diff
@@ -1564,6 +1564,8 @@ CREATE TABLE `invoice` (
   `description`         TEXT NOT NULL,
   `created_at`          TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
   `is_distributable`    TINYINT(1) NOT NULL ,
+  `is_credit_note`      TINYINT(1),
```
Why is this column in the `invoice` table?  Shouldn't it be in the `voucher` table?  Could it be a [voucher type](https://github.com/IMA-WorldHealth/bhima-2.X/issues/334) so that we do not introduce more columns for a single boolean?

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -13,10 +13,71 @@
+    );
+
+  return transaction.execute()
+    .then(function (rows) {
```
This `.then()` function doesn't do anything.  Please remove it.

@IMA-WorldHealth/local-contributors, this is a pattern I see a lot in code being submitted in PRs.  The following are equivalent:

``` js
return transaction.execute();
// ===
return transaction.execute()
  .then(function (rows) {
    return rows;
  });
```

In fact, you could do this over and over.

``` js
return $http.get('/some/route')
  .then(function (a) { return a; })
  .then(function (b) { return b; })
  .then((c) => { return c; })
  .then(d => d)
  .then((e) => e)
  .then(function (f) { return f; })
  // .. etc ...
  .then(function (z) { return z; });
```

All of the functions of the form `.then(function (x) { return x; });` do not add any value to the code and should be removed.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -13,10 +13,71 @@
+    `, [uid])
+
+    .addQuery(`
+      UPDATE invoice SET is_credit_note = 1, credit_note_by = ${user_id} 
```
Why do we need to do this?

Shouldn't the posting journal row have the `user_id` already?

lomamech replied:
I have done just because the table invoice table there was a user_id column that was allowing the user to know which generated the invoice and that's why I add the credit_note_by column for the user cancels an invoice

@jniles

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -13,10 +13,71 @@
 const db = require('../../../lib/db');
+const core = require('./core');
+
+var uuid = require('node-uuid');
```
Please use `const` for required libraries.  That way you do not accidentally overwrite your variables later in a function!

lomamech committed:
Improve and fixe Request for reverseInvoice

jniles commented on client/src/partials/accounts/templates/grid.creditNote.tmpl.html:
```diff
@@ -0,0 +1,13 @@
+  ng-repeat="(colRenderIndex, col) in colContainer.renderedColumns track by col.uid"
+  ui-grid-one-bind-id-grid="rowRenderIndex + '-' + col.uid + '-cell'"
+  class="ui-grid-cell"
+  ng-class="{ 'ui-grid-row-header-cell': col.isRowHeader, 'strike' : row.entity.is_credit_note === 1 }"
```
Cool!

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.js:
```diff
@@ -0,0 +1,36 @@
+  function submit(uuid) {
+    if (!uuid) { return; }
+
+    var journal = JournalVoucher.reverse(uuid);
```
This would probably be better called `promise`.

lomamech committed:
Sanitaze translate file en.json

lomamech committed:
Update and Reforme the way to be credit not with the using of table voucher and delete the property is_credit_note and credit_note_by

lomamech committed:
A unique identifier for insertion into the voucher table

lomamech committed:
Update forme for saved the description of Credit Note

lomamech committed:
Block the credit note without description

lomamech committed:
Resolve conflict with the branch Master

lomamech committed:
Rebase with the branch Master

lomamech committed:
Resolve Conflict

lomamech committed:
Continue to rebase with branch Master

lomamech committed:
Update, Improve and remake the integration and E2E test

jniles commented on client/src/i18n/fr.json:
```diff
@@ -822,20 +829,18 @@
+  "PURCHASES" : {
+    "PAGE_TITLE" : "Achats",
+    "PAYMENT_METHOD" : "Mode de paiement",
+    "PAYMENT_METHOD_EXPLANATION" : "Ce champ fournit les conditions de paiement de l'ordre d'achat convenu. Un achat peut être payée à l'avant, à la livraison, ou dans une autre combinaison de paiements.",
```
Very nice translations!

jniles commented on client/src/js/services/JournalVoucherService.js:
```diff
@@ -0,0 +1,28 @@
+ * @module JournalVoucherService
+ */
+
+function JournalVoucherService($http, util) {
```
There is already a service called `VoucherService` for creating journal vouchers.  You should you that service.

jniles commented on client/src/js/services/JournalVoucherService.js:
```diff
@@ -0,0 +1,28 @@
+  service.reverse = reverse;
+
+  /**
+   * This method facilitate annulling a transaction, 
```
This is a very nice description!

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.html:
```diff
@@ -0,0 +1,70 @@
+<form  name="ModalForm" ng-submit="ModalCtrl.submit(ModalCtrl.patientInvoice.uuid, ModalCtrl.creditNote)" novalidate>
```
This form should use the `bh-form-defaults` directive.

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.js:
```diff
@@ -0,0 +1,48 @@
+  var transferTypeId = typeId[0].id;
+
+
+  Invoices.read(vm.patientInvoice.uuid)    
```
This should use `Notify` to handle errors.

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.js:
```diff
@@ -0,0 +1,48 @@
+  function submit(uuid, creditNote) {
+    creditNote.type_id = transferTypeId;
+    
+    if (!uuid || !creditNote) { return; }
```
Should this give some sort of error message? How will the user know that nothing is happening?

jniles commented on client/src/partials/patient_invoice/registry/registry.js:
```diff
@@ -130,7 +132,18 @@ function InvoiceRegistryController(Invoices, Notify, Session, util, Receipt, App
+        Notify.success('FORM.INFOS.TRANSACTION_REVER_SUCCESS');
+        
+        return load();
+      });
```
Should this have error handling?

jniles commented on client/src/partials/patient_invoice/registry/registry.js:
```diff
@@ -130,7 +132,18 @@ function InvoiceRegistryController(Invoices, Notify, Session, util, Receipt, App
+  function creditNote(invoice) {
+    Invoices.openCreditNoteModal(invoice)
+      .then(function (data) {
+        // no parameters means the modal was dismissed.
```
I don't think that's true.  You call `$uibModalInstance.dismiss()` which will reject the modal promise.  If the modal is dismissed, this function is never called.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -13,10 +13,74 @@
+exports.reverse = reverse;
+
+
+// GET /Grade
```
This function is not mapped to `GET /Grade`.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -13,10 +13,74 @@
+
+
+// GET /Grade
+function lookReverseTransaction(uid, user_id, creditNote) {
```
Why is this called `lookReverseTransaction()` if it has an `INSERT` statement?  A better name would be `createReverseTransaction()` or something like that.

jniles commented on server/config/routes.js:
```diff
@@ -200,6 +200,7 @@ exports.configure = function configure(app) {
   // trial balance routes
   app.post('/journal/trialbalance', trialbalance.postTrialBalance);
   app.post('/journal/togeneralledger', trialbalance.postToGeneralLedger); // TODO : rename?
+  app.put('/journal/:uuid/reverse', journal.reverse);
```
This would be better as a POST method, since it is creating new records in the database, instead of modifying existing ones.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -49,3 +113,21 @@ function list(req, res, next) {
+ */
+function reverse(req, res, next) {
+  const uid = db.bid(req.params.uuid); 
+  var user_id = req.session.user.id;
```
Please use `camelCase` instead of `snake_case` for variable names in JS.

lomamech committed:
Remove the service JournalVoucherService instead voucherService

lomamech committed:
Improved Code
Using bh-form-defaults in modal Credit Note
Using Notify for catching Handle errors
Changing the way to track a valide form duiring submit credit note
Using createReverseTransaction instead lookReverseTransactin
Using Post instead Put for journal/uuid/reverse
Sanitaze code

lomamech committed:
Complete translate language

lomamech committed:
Complete translate file

lomamech committed:
Fix test E2E for registry.search and registry.spec

lomamech committed:
Fix and Sanitaze code

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.html:
```diff
@@ -0,0 +1,70 @@
+    </div>
+  </div>
+
+  <div class="modal-body" style="overflow: auto; max-height: 2000px;">
```
That's a very large max height!

In the future, it would be good to limit a modal's height to a part of the page height.  You can do it with the `vh` CSS unit.  [Here is the documentation](https://developer.mozilla.org/en/docs/Web/CSS/length#Viewport-percentage_lengths).  Like this:

``` html
<!-- this element will never be taller than 95% of the screen height!  Yay! -->
<element style="max-height: 95vh;">
</element>
```

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.html:
```diff
@@ -0,0 +1,70 @@
+<form  name="ModalForm" ng-submit="ModalCtrl.submit(ModalForm)" bh-form-defaults novalidate>
+  <div class="modal-header">
+    <div class="bhima-title">
```
In the future, do not use the class `bhima-title` in the modal headers.  @sfount has fixed modal header styling so that it displays a nice `headercrumb` without the `bhima-title` class.

jniles commented on client/src/partials/patient_invoice/registry/modalCreditNote.html:
```diff
@@ -0,0 +1,70 @@
+              <td>{{item.text}}</td>
+              <td>{{item.quantity}}</td>
+              <td>{{item.transaction_price | currency}}</td>
+              <td>{{item.quantity * item.transaction_price | currency}}</td>
```
This currency filter must have the currency of the invoice passed into it.

lomamech replied:
@jniles 
There is no currency id in the table invoice, it would be possible to use the company currency

jniles replied:
It would be good to use the enterprise currency.  What happens if it is in
FCs?  It will show USD for all the bills.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -49,3 +112,21 @@ function list(req, res, next) {
 }
+
+/**
+ * PUT /journal/:UUID/reverse
```
This comment should be adjusted to reflect the POST request.  This won't block the PR.

jniles commented on server/test/api/patientInvoiceCreditNote.js:
```diff
@@ -0,0 +1,30 @@
+    return agent.post('/journal/' + fetchableInvoiceUuid + '/reverse')
+      .send({ description : 'Credit Note', type_id: 9 })
+      .then(function (res) {
+        expect(res);
```
This assertion doesn't mean anything. :confused:

lomamech committed:
Sanitaze code adding Semicolon

lomamech committed:
Sanitaze test integration for Patient Credit Note

lomamech committed:
Sanitaze Comment in journal/index.js
